src/rules/interfaceNameRule.ts

```diff
@@ -47,60 +49,26 @@ export class Rule extends Lint.Rules.AbstractRule {
     public static FAILURE_STRING_NO_PREFIX = `interface name must not have an "I" prefix`;
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new NameWalker(sourceFile, this.getOptions()));
+        return this.applyWithFunction(sourceFile, walk, { never: this.ruleArguments.indexOf(OPTION_NEVER) !== -1 });
     }
 }
 
-class NameWalker extends Lint.RuleWalker {
-    public visitInterfaceDeclaration(node: ts.InterfaceDeclaration) {
-        const interfaceName = node.name.text;
-
-        const always = this.hasOption(OPTION_ALWAYS) || (this.getOptions() && this.getOptions().length === 0);
-
-        if (always) {
-            if (!this.startsWithI(interfaceName)) {
-                this.addFailureAtNode(node.name, Rule.FAILURE_STRING);
-            }
-        } else if (this.hasOption(OPTION_NEVER)) {
-            if (this.hasPrefixI(interfaceName)) {
-                this.addFailureAtNode(node.name, Rule.FAILURE_STRING_NO_PREFIX);
+function walk(ctx: Lint.WalkContext<{ never: boolean }>): void {
+    const { options: { never } } = ctx;
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (utils.isInterfaceDeclaration(node)) {
+            const { name } = node;
+            if (never && hasPrefixI(name.text)) {
+                ctx.addFailureAtNode(name, Rule.FAILURE_STRING_NO_PREFIX);
+            } else if (!never && name.text[0] !== "I") {
+                ctx.addFailureAtNode(name, Rule.FAILURE_STRING);
             }
         }
+        return ts.forEachChild(node, cb);
```

Comment:
nit: you could move this line to the else branch, because interfaces cannot occur inside other interfaces